File: desktop/src/main/workspace-gitignore.ts
import { readFile, writeFile } from "node:fs/promises";
import path from "node:path";

const AGENTHUB_RUNTIME_DIRECTORIES = [".agenthub", ".codex", ".claude", ".gemini"] as const;

export async function ensureWorkspaceGitIgnore(workspacePath: string): Promise<boolean> {
  const gitignorePath = path.join(workspacePath, ".gitignore");
  const current = await readOptionalText(gitignorePath);
  const next = mergeWorkspaceGitIgnore(current);
  if (next === current) {
    return false;
  }
  await writeFile(gitignorePath, next, "utf8");
  return true;
}

export function mergeWorkspaceGitIgnore(raw: string): string {
  const existingDirectories = new Set(
    raw
      .replace(/^\uFEFF/, "")
      .split(/\r?\n/)
      .map((line) => normalizeGitIgnorePattern(line))
      .filter((directory): directory is string => Boolean(directory)),
  );
  const missingEntries = AGENTHUB_RUNTIME_DIRECTORIES.filter((directory) => !existingDirectories.has(directory)).map(
    (directory) => `${directory}/`,
  );

  if (missingEntries.length === 0) {
    return raw;
  }

  const normalized = raw.replace(/\r\n/g, "\n").replace(/\r/g, "\n").trimEnd();
  const prefix = normalized ? `${normalized}\n` : "";
  return `${prefix}${missingEntries.join("\n")}\n`;
}

function normalizeGitIgnorePattern(line: string): string | null {
  const trimmed = line.trim();
  if (!trimmed || trimmed.startsWith("#") || trimmed.startsWith("!")) {
    return null;
  }

  const pattern = trimmed
    .replace(/\s+#.*$/, "")
    .replace(/\\/g, "/")
    .replace(/^\.\//, "")
    .replace(/^\//, "")
    .replace(/\/\*\*?$/, "")
    .replace(/\/+$/, "");
  return AGENTHUB_RUNTIME_DIRECTORIES.includes(pattern as (typeof AGENTHUB_RUNTIME_DIRECTORIES)[number])
    ? pattern
    : null;
}

async function readOptionalText(filePath: string): Promise<string> {
  try {
    return await readFile(filePath, "utf8");
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      return "";
    }
    throw error;
  }
}

File: native/AgentHub.Native/src/AgentHub.Native.Core/Workspaces/WorkspaceGitIgnore.cs
namespace AgentHub.Native.Core.Workspaces;

public static class WorkspaceGitIgnore
{
    private static readonly string[] RuntimeDirectories = [".agenthub", ".codex", ".claude", ".gemini"];

    public static async Task<bool> EnsureAsync(
        string workspacePath,
        CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(workspacePath);
        var gitIgnorePath = Path.Combine(workspacePath, ".gitignore");
        var current = File.Exists(gitIgnorePath)
            ? await File.ReadAllTextAsync(gitIgnorePath, cancellationToken).ConfigureAwait(false)
            : "";
        var next = Merge(current);
        if (next == current)
        {
            return false;
        }

        Directory.CreateDirectory(workspacePath);
        await File.WriteAllTextAsync(gitIgnorePath, next, cancellationToken).ConfigureAwait(false);
        return true;
    }

    public static string Merge(string raw)
    {
        var existing = raw
            .TrimStart('\uFEFF')
            .Replace("\r\n", "\n", StringComparison.Ordinal)
            .Replace('\r', '\n')
            .Split('\n')
            .Select(NormalizePattern)
            .OfType<string>()
            .ToHashSet(StringComparer.Ordinal);
        var missing = RuntimeDirectories
            .Where(directory => !existing.Contains(directory))
            .Select(directory => $"{directory}/")
            .ToArray();

        if (missing.Length == 0)
        {
            return raw;
        }

        var normalized = raw
            .Replace("\r\n", "\n", StringComparison.Ordinal)
            .Replace('\r', '\n')
            .TrimEnd();
        var prefix = string.IsNullOrEmpty(normalized) ? "" : $"{normalized}\n";
        return $"{prefix}{string.Join('\n', missing)}\n";
    }

    private static string? NormalizePattern(string line)
    {
        var trimmed = line.Trim();
        if (string.IsNullOrEmpty(trimmed) || trimmed.StartsWith('#') || trimmed.StartsWith('!'))
        {
            return null;
        }

        var pattern = trimmed;
        var commentIndex = pattern.IndexOf(" #", StringComparison.Ordinal);
        if (commentIndex >= 0)
        {
            pattern = pattern[..commentIndex].TrimEnd();
        }

        pattern = pattern.Replace('\\', '/');
        if (pattern.StartsWith("./", StringComparison.Ordinal))
        {
            pattern = pattern[2..];
        }

        pattern = pattern.TrimStart('/');
        pattern = pattern.TrimEnd('/');
        if (pattern.EndsWith("/**", StringComparison.Ordinal))
        {
            pattern = pattern[..^3].TrimEnd('/');
        }
        else if (pattern.EndsWith("/*", StringComparison.Ordinal))
        {
            pattern = pattern[..^2].TrimEnd('/');
        }

        if (!pattern.StartsWith('.'))
        {
            pattern = $".{pattern}";
        }

        return RuntimeDirectories.Contains(pattern, StringComparer.Ordinal) ? pattern : null;
    }
}
